Zend/zend_vm_def.h

```diff
@@ -9544,6 +9544,136 @@ ZEND_VM_HANDLER(202, ZEND_CALLABLE_CONVERT, UNUSED, UNUSED)
 }
 
+ZEND_VM_HANDLER(204, ZEND_FRAMELESS_ICALL_0, UNUSED, UNUSED)
+{
+	USE_OPLINE
+	SAVE_OPLINE();
+	zend_frameless_function_0 function = (zend_frameless_function_0)ZEND_FLF_HANDLER(opline);
+	ZVAL_NULL(EX_VAR(opline->result.var));
+	function(EX_VAR(opline->result.var));
+	if (EG(exception)) {
+		HANDLE_EXCEPTION();
+	}
+	ZEND_VM_NEXT_OPCODE();
+}
+
+ZEND_VM_HANDLER(205, ZEND_FRAMELESS_ICALL_1, CONST|TMP|VAR|CV, UNUSED)
+{
+	USE_OPLINE
+	SAVE_OPLINE();
+	zend_frameless_function_1 function = (zend_frameless_function_1)ZEND_FLF_HANDLER(opline);
+	ZVAL_NULL(EX_VAR(opline->result.var));
+
+	zval *arg1;
+	if (OP1_TYPE == IS_CV) {
+		/* Inlined GET_OPx_ZVAL_PTR_DEREF to nest the EG(exception) check. */
+		arg1 = EX_VAR(opline->op1.var);
+		if (UNEXPECTED(Z_TYPE_P(arg1) == IS_UNDEF)) {
+			arg1 = zval_undefined_cv(opline->op1.var EXECUTE_DATA_CC);
+			if (EG(exception)) {
+				FREE_OP1();
+				HANDLE_EXCEPTION();
+			}
+		}
+		ZVAL_DEREF(arg1);
+	} else {
+		arg1 = GET_OP1_ZVAL_PTR_DEREF(BP_VAR_R);
+	}
+
+	function(EX_VAR(opline->result.var), arg1);
+	FREE_OP1();
+	if (EG(exception)) {
+		HANDLE_EXCEPTION();
+	}
+	ZEND_VM_NEXT_OPCODE();
+}
+
+ZEND_VM_HANDLER(206, ZEND_FRAMELESS_ICALL_2, CONST|TMP|VAR|CV, CONST|TMP|VAR|CV)
+{
+	USE_OPLINE
+	SAVE_OPLINE();
+	zend_frameless_function_2 function = (zend_frameless_function_2)ZEND_FLF_HANDLER(opline);
+	ZVAL_NULL(EX_VAR(opline->result.var));
+
+	zval *arg1;
+	if (OP1_TYPE == IS_CV) {
+		/* Inlined GET_OPx_ZVAL_PTR_DEREF to nest the EG(exception) check. */
+		arg1 = EX_VAR(opline->op1.var);
+		if (UNEXPECTED(Z_TYPE_P(arg1) == IS_UNDEF)) {
+			arg1 = zval_undefined_cv(opline->op1.var EXECUTE_DATA_CC);
+			if (EG(exception)) {
+				FREE_OP1();
+				FREE_OP2();
+				HANDLE_EXCEPTION();
+			}
+		}
+		ZVAL_DEREF(arg1);
+	} else {
+		arg1 = GET_OP1_ZVAL_PTR_DEREF(BP_VAR_R);
+	}
+
+	zval *arg2;
+	if (OP2_TYPE == IS_CV) {
+		/* Inlined GET_OPx_ZVAL_PTR_DEREF to nest the EG(exception) check. */
+		arg2 = EX_VAR(opline->op2.var);
+		if (UNEXPECTED(Z_TYPE_P(arg2) == IS_UNDEF)) {
+			arg2 = zval_undefined_cv(opline->op2.var EXECUTE_DATA_CC);
+			if (EG(exception)) {
+				FREE_OP1();
+				FREE_OP2();
+				HANDLE_EXCEPTION();
+			}
+		}
+		ZVAL_DEREF(arg2);
+	} else {
+		arg2 = GET_OP2_ZVAL_PTR_DEREF(BP_VAR_R);
+	}
+
+	function(EX_VAR(opline->result.var), arg1, arg2);
+	FREE_OP1();
+	/* Set OP1 to UNDEF in case FREE_OP2() throws. */
+	if ((OP1_TYPE & (IS_VAR|IS_TMP_VAR)) != 0 && (OP2_TYPE & (IS_VAR|IS_TMP_VAR)) != 0) {
+		ZVAL_UNDEF(EX_VAR(opline->op1.var));
+	}
+	FREE_OP2();
+	if (EG(exception)) {
+		HANDLE_EXCEPTION();
+	}
+	ZEND_VM_NEXT_OPCODE();
+}
+
+ZEND_VM_HANDLER(207, ZEND_FRAMELESS_ICALL_3, ANY, ANY)
+{
+	USE_OPLINE
+	SAVE_OPLINE();
+	zend_frameless_function_3 function = (zend_frameless_function_3)ZEND_FLF_HANDLER(opline);
+	ZVAL_NULL(EX_VAR(opline->result.var));
+	zval *arg1 = GET_OP1_ZVAL_PTR_DEREF(BP_VAR_R);
+	zval *arg2 = GET_OP2_ZVAL_PTR_DEREF(BP_VAR_R);
+	zval *arg3 = GET_OP_DATA_ZVAL_PTR_DEREF(BP_VAR_R);
```

Comment:
Exceptions in case of undefined CV are not handled (like in ZEND_FRAMELESS_ICALL_2),

Reply:
Note that `ZEND_FRAMELESS_ICALL_2` contains an optimization. I didn't apply this to `ZEND_FRAMELESS_ICALL_3` because it was unspecialized, and otherwise would have had to inline all of `get_zval_ptr_deref`. `GET_OPx_ZVAL_PTR_DEREF` already handles undefined CVs, so I think the code *should* be correct. Technically there's still a small deviation from normal function calls because we check for all undefined vars at once without checking for exceptions, but I accepted that for the sake of performance.

If we do add specializations I can add the same optimization as in `ZEND_FRAMELESS_ICALL_2`.

Reply:
You are right!